ext/re2/extconf.rb

```diff
@@ -9,370 +9,278 @@
-  host.gsub(/i386/, "i686")
-end
+    def build_with_system_libraries
+      header_dirs = [
+        "/usr/local/include",
+        "/opt/homebrew/include",
+        "/usr/include"
+      ]
 
-def target_arch
-  RbConfig::CONFIG['arch']
-end
+      lib_dirs = [
+        "/usr/local/lib",
+        "/opt/homebrew/lib",
+        "/usr/lib"
+      ]
 
-def with_temp_dir
-  Dir.mktmpdir do |temp_dir|
-    Dir.chdir(temp_dir) do
-      yield
+      dir_config("re2", header_dirs, lib_dirs)
     end
-  end
-end
 
-#
-#  main
-#
-do_help if arg_config('--help')
-
-if ENV["CC"]
-  RbConfig::MAKEFILE_CONFIG["CC"] = ENV["CC"]
-  RbConfig::CONFIG["CC"] = ENV["CC"]
-end
-
-if ENV["CXX"]
-  RbConfig::MAKEFILE_CONFIG["CXX"] = ENV["CXX"]
-  RbConfig::CONFIG["CXX"] = ENV["CXX"]
-end
+    def build_with_vendored_libraries
+      message "Building re2 using packaged libraries.\n"
 
-def build_extension(static_p = false)
-  # Enable optional warnings but disable deprecated register warning for Ruby 2.6 support
-  $CFLAGS << " -Wall -Wextra -funroll-loops"
-  $CPPFLAGS << " -Wno-register"
+      abseil_recipe, re2_recipe = load_recipes
 
-  # Pass -x c++ to force gcc to compile the test program
-  # as C++ (as it will end in .c by default).
-  compile_options = "-x c++"
+      process_recipe(abseil_recipe) do |recipe|
+        recipe.configure_options += ['-DABSL_PROPAGATE_CXX_STD=ON', '-DCMAKE_CXX_VISIBILITY_PRESET=hidden']
+        # Workaround for https://github.com/abseil/abseil-cpp/issues/1510
+        recipe.configure_options += ['-DCMAKE_CXX_FLAGS=-DABSL_FORCE_WAITER_MODE=4'] if windows?
+      end
 
-  have_library("stdc++")
-  have_header("stdint.h")
-  have_func("rb_gc_mark_movable") # introduced in Ruby 2.7
+      abseil_recipe.activate
 
-  if !static_p and !have_library("re2")
-    abort "You must have re2 installed and specified with --with-re2-dir, please see https://github.com/google/re2/wiki/Install"
-  end
+      process_recipe(re2_recipe) do |recipe|
+        recipe.configure_options += ["-DCMAKE_PREFIX_PATH=#{abseil_recipe.path}", '-DCMAKE_CXX_FLAGS=-DNDEBUG',
+                                     '-DCMAKE_CXX_VISIBILITY_PRESET=hidden']
+      end
 
-  minimal_program = <<SRC
-#include <re2/re2.h>
-int main() { return 0; }
-SRC
+      dir_config("re2", File.join(re2_recipe.path, 'include'), File.join(re2_recipe.path, 'lib'))
+      dir_config("abseil", File.join(abseil_recipe.path, 'include'), File.join(abseil_recipe.path, 'lib'))
 
-  re2_requires_version_flag = checking_for("re2 that requires explicit C++ version flag") do
-    !try_compile(minimal_program, compile_options)
-  end
+      pkg_config_paths = [
+        "#{abseil_recipe.path}/lib/pkgconfig",
+        "#{re2_recipe.path}/lib/pkgconfig"
+      ]
+      pkg_config_paths.prepend(ENV['PKG_CONFIG_PATH']) if ENV['PKG_CONFIG_PATH']
+      ENV['PKG_CONFIG_PATH'] = pkg_config_paths.join(File::PATH_SEPARATOR)
```

Comment:
I hoped we could delete this but the gem doesn’t compile correctly without it when using the new `mkmf_config` feature even though sqlite3-ruby doesn’t seem to set it (see https://github.com/sparklemotion/sqlite3-ruby/pull/405/files). @stanhu what’s different about our setup?

Reply:
Perhaps we should be using `mkmf_config` for Abseil too ala [Nokogiri and its dependencies](https://github.com/sparklemotion/nokogiri/pull/2977/files)?

Reply:
I think there is still a risk of linking against system shared libraries instead of the build shared libraries; if that's a concern please don't migrate to mkmf_config yet. And please don't use it unless you're comfortable with the API or the behavior changing (it's still "experimental").

Reply:
Looking at https://github.com/flavorjones/mini_portile/blob/52fb0bc41c89a10f1ac7b5abcf0157e059194374/lib/mini_portile2/mini_portile.rb#L272-L280 and comparing it to our own logic:
 https://github.com/mudge/re2/blob/cc4940a3ff7c1ca2c2b00ddc39bda1826d495d9c/ext/re2/extconf.rb#L283-L296

I _think_ we're doing very similar things (expanding `-l` into full static paths, ensuring `-L` flags are prepended) however one thing I'm not sure on is how it'll handle `re2.pc` including Abseil's libraries:

```
$ pkg-config --libs --static libre2/2024-03-01/lib/pkgconfig/re2.pc
-L/Users/mudge/Projects/re2/ports/arm64-apple-darwin22/libre2/2024-03-01/lib -L/opt/homebrew/Cellar/abseil/20230802.1/lib -pthread -lre2 -labsl_flags -labsl_flags_internal -labsl_flags_marshalling -labsl_flags_reflection -labsl_flags_private_handle_accessor -labsl_flags_commandlineflag -labsl_flags_commandlineflag_internal -labsl_flags_config -labsl_flags_program_name -labsl_cord -labsl_cordz_info -labsl_cord_internal -labsl_cordz_functions -labsl_cordz_handle -labsl_crc_cord_state -labsl_crc32c -labsl_crc_internal -labsl_crc_cpu_detect -labsl_raw_hash_set -labsl_hash -labsl_city -labsl_bad_variant_access -labsl_low_level_hash -labsl_hashtablez_sampler -labsl_exponential_biased -labsl_bad_optional_access -labsl_str_format_internal -labsl_synchronization -labsl_graphcycles_internal -labsl_kernel_timeout_internal -labsl_stacktrace -labsl_symbolize -labsl_debugging_internal -labsl_demangle_internal -labsl_malloc_internal -labsl_time -labsl_civil_time -labsl_strings -labsl_string_view -labsl_strings_internal -labsl_base -labsl_spinlock_wait -labsl_int128 -labsl_throw_delegate -labsl_raw_logging_internal -labsl_log_severity -labsl_time_zone
```

Reply:
@mudge I think what's different here is that the `re2` gem needs to pull in `.pc` files for two libraries that are built: `libre2` and `nokogiri`. I think the `sqlite` gem only needs the the `sqlite3` library.

Could we make `mkmf_config` accept an array of `pkg` values?

Reply:
I suspect things are complicated by Abseil having over 180 separate `.pc` files so there’s no simple single configuration for that. However, `re2.pc` refers to the exact ones we need but I can’t tell from MiniPortile2’s source if it supports resolving static references to _other_ packages outside `libre2`.

Reply:
I can confirm that setting `PKG_CONFIG_PATH` to include both libraries' `lib/pkgconfig` directories fixes Abseil's `-L` flag from using my system install to the vendored one:

```
$ env PKG_CONFIG_PATH=/Users/mudge/Projects/re2/ports/arm64-apple-darwin22/libre2/2024-03-01/lib/pkgconfig:/Users/mudge/Projects/re2/ports/arm64-apple-darwin22/abseil/20240116.1/lib/pkgconfig pkg-config --libs --static re2.pc
-L/Users/mudge/Projects/re2/ports/arm64-apple-darwin22/libre2/2024-03-01/lib -L/Users/mudge/Projects/re2/ports/arm64-apple-darwin22/abseil/20240116.1/lib -pthread -lre2 -labsl_flags_internal -labsl_flags_marshalling -labsl_flags_reflection -labsl_flags_private_handle_accessor -labsl_flags_commandlineflag -labsl_flags_commandlineflag_internal -labsl_flags_config -labsl_flags_program_name -labsl_cord -labsl_cordz_info -labsl_cord_internal -labsl_cordz_functions -labsl_cordz_handle -labsl_crc_cord_state -labsl_crc32c -labsl_crc_internal -labsl_crc_cpu_detect -labsl_raw_hash_set -labsl_hash -labsl_city -labsl_bad_variant_access -labsl_low_level_hash -labsl_hashtablez_sampler -labsl_exponential_biased -labsl_bad_optional_access -labsl_str_format_internal -labsl_synchronization -labsl_graphcycles_internal -labsl_kernel_timeout_internal -labsl_stacktrace -labsl_symbolize -labsl_debugging_internal -labsl_demangle_internal -labsl_malloc_internal -labsl_time -labsl_civil_time -labsl_strings -labsl_strings_internal -labsl_string_view -labsl_base -labsl_spinlock_wait -labsl_int128 -labsl_throw_delegate -labsl_raw_logging_internal -labsl_log_severity -labsl_time_zone
```

Ideally, MiniPortile2 would then try to resolve every subsequent `-l` flag against _both_ lib directories but as it is, I think [it always resolves them relative to the output of `pkg-config variable=libdir`](https://github.com/flavorjones/mini_portile/blob/52fb0bc41c89a10f1ac7b5abcf0157e059194374/lib/mini_portile2/mini_portile.rb#L345) which is only RE2's lib directory and doesn't include Abseil's.

@flavorjones would it make sense to suppose parsing the output of `pkg-config --libs-only-L --static` and checking for each `-l` flag against those directories in turn?

e.g. `pkg-config --libs-only-L --static re2.pc` returns `-L/Users/mudge/Projects/re2/ports/arm64-apple-darwin22/libre2/2024-03-01/lib -L/Users/mudge/Projects/re2/ports/arm64-apple-darwin22/abseil/20240116.1/lib`, `pkg-config --libs-only-l --static re2.pc` returns (among other things) `-lre2 -labsl_flags_internal` so we'd try to find `libre2.a` in both `/Users/mudge/Projects/re2/ports/arm64-apple-darwin22/libre2/2024-03-01/lib` and `/Users/mudge/Projects/re2/ports/arm64-apple-darwin22/abseil/20240116.1/lib` and find it in the first one. Then we'd try to find `libabsl_flags_internal.a` and find it in the latter.

Reply:
@mudge I'm definitely open to doing that work if it would make linking more reliable and easier for gem maintainers. Would you mind terribly opening an issue at https://github.com/flavorjones/mini_portile/issues that describes a test scenario that I can code up and use to drive the implementation?

Reply:
I’ve raised https://github.com/flavorjones/mini_portile/issues/138 but please let me know if you need more detail or a more specific test case.